File: packages/option-picker/ios/Plugin/OptionPickerPlugin.swift
import Foundation
import Capacitor

@objc(OptionPickerPlugin)
public class OptionPickerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "OptionPickerPlugin"
    public let jsName = "OptionPicker"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "present", returnType: CAPPluginReturnPromise)
    ]
    public static let tag = "OptionPickerPlugin"

    private var implementation: OptionPicker?

    override public func load() {
        self.implementation = OptionPicker(plugin: self)
    }

    @objc func present(_ call: CAPPluginCall) {
        do {
            let options = try PresentOptions(call)
            implementation?.present(options) { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                    return
                }
                self.resolveCall(call, result)
            }
        } catch {
            rejectCall(call, error)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", OptionPickerPlugin.tag, "] ", error)
        let code = (error as? CustomError)?.code
        call.reject(error.localizedDescription, code)
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}

File: packages/option-picker/android/src/main/java/io/capawesome/capacitorjs/plugins/optionpicker/OptionPickerPlugin.java
package io.capawesome.capacitorjs.plugins.optionpicker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.optionpicker.classes.CustomException;
import io.capawesome.capacitorjs.plugins.optionpicker.classes.options.PresentOptions;
import io.capawesome.capacitorjs.plugins.optionpicker.classes.results.PresentResult;
import io.capawesome.capacitorjs.plugins.optionpicker.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.optionpicker.interfaces.Result;

@CapacitorPlugin(name = "OptionPicker")
public class OptionPickerPlugin extends Plugin {

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String TAG = "OptionPickerPlugin";

    private OptionPicker implementation;

    @Override
    public void load() {
        super.load();
        this.implementation = new OptionPicker(this);
    }

    @PluginMethod
    public void present(PluginCall call) {
        try {
            PresentOptions options = new PresentOptions(call);
            NonEmptyResultCallback<PresentResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull PresentResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.present(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        String code = null;
        if (exception instanceof CustomException) {
            code = ((CustomException) exception).getCode();
        }
        Logger.error(TAG, message, exception);
        call.reject(message, code);
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}
